### PR TITLE
feat(misc): create `pyproject.toml`

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Reference: #2817 

# What does this PR do?

This PR aims to minimize the number of configuration files for the project. Instead of a separate `.style.yapf`, `mypy.ini`, `pytest.ini`,`pylintrc`,`pytype.cfg` and `setup.cfg`, we now would have all the configurations within the `pyproject.toml` file leading to a overall minimal code structure. I've also updated the `.pre-commit-config.yaml` file with the addition of a "check-toml" hook.

### CODE REVIEW

**Comment by cgarciae** [comment]
Hey @SauravMaheshkar, have a look at the `tests/run_all_tests.sh` file, you might need to fix it based on these changes.

**Comment by SauravMaheshkar** [comment]
Addressed in [735bde7](https://github.com/google/flax/pull/2828/commits/735bde7605dc40b70bc781b1a0349fee24c1cefe)

**Review by levskaya — APPROVED** [review]
Thanks for the help, I've been wanting to unify things for a while.

**Comment by levskaya** [comment]
Hmm @SauravMaheshkar it looks like mypy is freaking out?  Are its configs not quite setup as they were before?

**Comment by SauravMaheshkar** [comment]
Maybe a43ce84871f7bd553322ef1e2dcc3a3b58dd0d88 will work, I separated the `mypy` ignores and tried running the following command locally

```lang-bash
tests/run_all_tests.sh --no-doctest --no-pytest --no-pytype --use-venv
```

**Comment by cgarciae** [comment]
If this lands we can try getting rid of `setup.py` 🙂

File: .gitignore [hunk 1]
```diff
@@ -12,4 +12,7 @@ build/
 .pytype
 .vscode/*
 /.devcontainer
-docs/**/tmp
\ No newline at end of file
+docs/**/tmp
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply .gitignore:15]:
```
venv/ # not necessary
.pytest_cache/ # not necessary
.mypy_cache/   # not necessary
```

[the remark this answers is not in the document]
Comment (SauravMaheshkar) [reply .gitignore:15]: Okay I'll remove them.

**Comment by SauravMaheshkar** [comment]
@cgarciae any idea as to why the doctest is failing ? I have not made any changes to the `doctest` configuration

**Comment by cgarciae** [comment]
@SauravMaheshkar seems like a slight numerical inaccuracy, not sure what this has to do with the proposed changes. Maybe to rebase/merge with the latest changes in `main`.

**Pull request closed without merging**